dev/src/bulk-writer.ts

```diff
@@ -229,56 +235,62 @@ class BulkCommitBatch {
         results = await this.writeBatch.bulkCommit();
       } catch (err) {
         // Map the failure to each individual write's result.
-        results = [...this.pendingOps.keys()].map(path => {
-          return {key: path, writeTime: null, status: wrapError(err, stack)};
+        results = this.pendingOps.map(op => {
+          return {key: op.key, writeTime: null, status: wrapError(err, stack)};
         });
       }
       this.processResults(results);
 
-      if (this.pendingOps.size > 0) {
+      if (this.pendingOps.length > 0) {
         logger(
           'BulkWriter.bulkCommit',
           null,
           `Current batch failed at retry #${attempt}. Num failures: ` +
-            `${this.pendingOps.size}.`
+            `${this.pendingOps.length}.`
         );
 
-        this.writeBatch = new WriteBatch(this.firestore, this.writeBatch, [
-          ...this.pendingOps.keys(),
-        ]);
+        this.writeBatch = new WriteBatch(
+          this.firestore,
+          this.writeBatch,
+          this.pendingOps.map(op => op.key)
+        );
       } else {
         this.completedDeferred.resolve();
         return;
       }
     }
 
-    this.failRemainingOperations(results);
+    this.processResults(results, /* failRemainingOperations= */ true);
     this.completedDeferred.resolve();
   }
 
   /**
    * Resolves the individual operations in the batch with the results.
    */
-  private processResults(results: BatchWriteResult[]): void {
-    for (const result of results) {
-      if (result.status.code === Status.OK) {
-        this.pendingOps.get(result.key)!.resolve(result);
-        this.pendingOps.delete(result.key);
+  private processResults(
+    results: BatchWriteResult[],
+    failRemainingOperations = false
+  ): void {
+    const resolvedOps = [];
+    for (let i = 0; i < results.length; i++) {
+      const result = results[i];
+      if (failRemainingOperations) {
+        assert(
+          result.status.code !== Status.OK,
+          'Should not fail successful operation'
+        );
+        this.pendingOps[i].result.reject(result.status);
+        resolvedOps.push(i);
+      } else if (result.status.code === Status.OK) {
+        this.pendingOps[i].result.resolve(result);
+        resolvedOps.push(i);
       } else if (!this.shouldRetry(result.status.code)) {
-        this.pendingOps.get(result.key)!.reject(result.status);
-        this.pendingOps.delete(result.key);
+        this.pendingOps[i].result.reject(result.status);
+        resolvedOps.push(i);
       }
     }
-  }
-
-  private failRemainingOperations(results: BatchWriteResult[]): void {
-    for (const result of results) {
-      assert(
-        result.status.code !== Status.OK,
-        'Should not fail successful operation'
-      );
-      this.pendingOps.get(result.key)!.reject(result.status);
-      this.pendingOps.delete(result.key);
+    for (let i = resolvedOps.length - 1; i >= 0; i--) {
+      this.pendingOps.splice(resolvedOps[i], 1);
```

Comment:
Aren't these indices off after the first splice? I wonder if we can simply build up the reverse array in the loop above and replace this.pendingOps with it.

Reply:
Since we're going in reverse order and deleting the last element in the array each time, the remaining indices won't be affected.

Reply:
Can you add a comment?

Reply:
done.